### PR TITLE
ref: add sentry pypi as an extra-index-url

### PR DESCRIPTION
**Pull request by asottile-sentry**
this is an incremental step towards full adoption of internal pypi

### CODE REVIEW

File: scripts/lib.sh [hunk 1]
```diff
@@ -91,8 +95,7 @@ sudo-askpass() {
 }
 
 upgrade-pip() {
-    grep -E '^(pip|setuptools|wheel)==' requirements-dev-frozen.txt |
-        xargs pip install --upgrade
+    pip-install $(grep -E '^(pip|setuptools|wheel)==' requirements-dev-frozen.txt)
```
Comment (armenzg) [inline scripts/lib.sh:98]: Let me see if I understand this correctly. One command extracts the indexes to use and the other one filters the 3 packages to pin. This all becomes a single command as per below.

Should the function be called something more clarifying? (e.g. `install-pinned-pip-tools`)

```
❯ pip install $(grep ^-- requirements-base.txt) $(grep -E '^(pip|setuptools|wheel)==' requirements-dev-frozen.txt)
Looking in indexes: https://pypi.devinfra.sentry.io/simple, https://pypi.org/simple
Requirement already satisfied: wheel==0.37.1 in ./.venv/lib/python3.8/site-packages (0.37.1)
Requirement already satisfied: pip==22.1.2 in ./.venv/lib/python3.8/site-packages (22.1.2)
Requirement already satisfied: setuptools==56.0.0 in ./.venv/lib/python3.8/site-packages (56.0.0)
```

File: scripts/lib.sh [hunk 2]
```diff
@@ -103,10 +106,6 @@ install-py-dev() {
 
     echo "--> Installing Sentry (for development)"
     if query-apple-m1; then
-        # This installs pyscopg-binary2 since there's no arm64 wheel
-        # This saves having to install postgresql on the Developer's machine + using flags
-        # https://github.com/psycopg/psycopg2/issues/1286
-        pip install https://storage.googleapis.com/python-arm64-wheels/psycopg2_binary-2.8.6-cp38-cp38-macosx_11_0_arm64.whl
```
Comment (armenzg) [inline scripts/lib.sh:109]: Boom!

**Review by armenzg — APPROVED** [review]
Testing:
* `make install-py-dev` runs successfully
* `make freeze-requirements` runs successfully on a M1
  * It takes a really long time (almost 2 minutes). I see a lot of `clang` activity
  * No diff changes; has anyone run this on a non-M1 laptop?
* `rm -rf .venv; direnv allow` runs successfully
   * It took 1m51secs
   * I'm surprised to see some new packages being downloaded. I would have expected everything already cached if we have been using freezing for a while
   * A run of `make install-py-dev` took 24sec after that
   
```text
Downloading https://pypi.devinfra.sentry.io/wheels/PyYAML-5.4-cp38-cp38-macosx_12_0_arm64.whl (171 kB)
Downloading https://pypi.devinfra.sentry.io/wheels/lxml-4.6.5-cp38-cp38-macosx_12_0_arm64.whl (2.6 MB)
```

File: requirements-dev-only-frozen.txt [hunk 3]
```diff
@@ -21,7 +24,7 @@ freezegun==1.1.0
 honcho==1.0.0
 identify==2.5.1
 idna==3.3
-importlib-resources==5.8.0
+importlib-resources==5.9.0
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply requirements-dev-only-frozen.txt:27]: ah I wanted to trigger getsentry to make sure that worked -- I also accidentally _removed_ this by freezing against python 3.10 first so this was an accidental coincidence here

File: scripts/lib.sh [hunk 4]
```diff
@@ -18,6 +18,10 @@ fi
 
 venv_name=".venv"
 
+pip-install() {
+    pip install $(grep ^-- requirements-base.txt) "$@"
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply scripts/lib.sh:22]: eheh yeah this is a little trick -- it's finding any lines that start with `--` in the requirements file (so it gets the index-url / extra-index-url arguments)

File: scripts/lib.sh (hunk 1 above)
Comment (asottile-sentry) [reply scripts/lib.sh:98]: yep!  -- had to change the syntax here a little bit because you can't xargs a function

**Comment by asottile-sentry** [comment]
> Testing:
> 
>     * `make install-py-dev` runs successfully
> 
>     * `make freeze-requirements` runs successfully on a M1
>       
>       * It takes a really long time (almost 2 minutes). I see a lot of `clang` activity
>       * No diff changes; has anyone run this on a non-M1 laptop?

yeah the "takes a long time" is because we haven't put `grpcio` wheels for M1 on pypi yet.  I've run it on linux but haven't run it on non-M1

> 
>     * `rm -rf .venv; direnv allow` runs successfully
>       
>       * It took 1m51secs
>       * I'm surprised to see some new packages being downloaded. I would have expected everything already cached if we have been using freezing for a while

for anything that we were previously installing from public pypi (and are now installing from internal pypi) they wouldn't have been cached

>       * A run of `make install-py-dev` took 24sec after that
> 
> 
> ```
> Downloading https://pypi.devinfra.sentry.io/wheels/PyYAML-5.4-cp38-cp38-macosx_12_0_arm64.whl (171 kB)
> Downloading https://pypi.devinfra.sentry.io/wheels/lxml-4.6.5-cp38-cp38-macosx_12_0_arm64.whl (2.6 MB)
> ```

**Comment by asottile-sentry** [comment]
superseded by #38255

**Pull request closed without merging**